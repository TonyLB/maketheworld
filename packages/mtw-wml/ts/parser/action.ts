import { ConverterMixinFactory, isTypedParseTagOpen } from "../convert/functionMixins"
import { ParseTagFactory, ParseActionTag, ParseCommentTag } from "./baseClasses"
import { validateProperties, ExtractProperties, validateContents } from "./utils"

export const parseActionFactory: ParseTagFactory<ParseActionTag> = ({ open, contents, endTagToken }) => {
    const validate = validateProperties<ExtractProperties<ParseActionTag, never>>({
        open,
        endTagToken,
        required: {
            key: ['key'],
            src: ['expression']
        }
    })
    validateContents<ParseCommentTag>({
        contents,
        legalTags: [],
        ignoreTags: ['Whitespace', 'Comment']
    })
    return {
        type: 'Tag',
        tag: {
            ...validate,
            tag: 'Action',
            startTagToken: open.startTagToken,
            endTagToken
        }
    }    
}

export const ParseActionMixin = ConverterMixinFactory({
    typeGuard: isTypedParseTagOpen('Action'),
    convert: ({ open, contents, endTagToken }) => {
        const validate = validateProperties<ExtractProperties<ParseActionTag, never>>({
            open,
            endTagToken,
            required: {
                key: ['key'],
                src: ['expression']
            }
        })
        validateContents<ParseCommentTag>({
            contents,
            legalTags: [],
            ignoreTags: ['Whitespace', 'Comment']
        })
        return {
            type: 'Tag',
            tag: {
                ...validate,
                tag: 'Action',
                startTagToken: open.startTagToken,
                endTagToken
            }
        }    
    }
})

export default parseActionFactory
