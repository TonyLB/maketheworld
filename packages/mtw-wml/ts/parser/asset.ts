import { ParseTagFactory, ParseAssetTag, ParseStoryTag, ParseAssetLegalContents } from "./baseClasses"
import { validateProperties, validateContents, ExtractProperties } from "./utils"

export const parseAssetFactory: ParseTagFactory<ParseAssetTag> = ({ open, contents, endTagToken }) => {
    const validate = validateProperties<ExtractProperties<ParseAssetTag, never>>({
        open,
        endTagToken,
        required: {
            key: ['key']
        },
        optional: {
            fileName: ['literal'],
            zone: ['literal'],
            subFolder: ['literal'],
            player: ['literal']
        }
    })
    const parseContents = validateContents<ParseAssetLegalContents>({
        contents,
        legalTags: ['Action', 'Computed', 'If', 'Else', 'ElseIf', 'Exit', 'Feature', 'Bookmark', 'Import', 'Room', 'Variable', 'Map'],
        ignoreTags: ['Whitespace', 'Comment']
    })
    return {
        type: 'Tag',
        tag: {
            ...validate,
            tag: 'Asset',
            startTagToken: open.startTagToken,
            endTagToken,
            contents: parseContents
        }
    }    
}

export const parseStoryFactory: ParseTagFactory<ParseStoryTag> = ({ open, contents, endTagToken }) => {
    const validate = validateProperties<ExtractProperties<ParseStoryTag, never>>({
        open,
        endTagToken,
        required: {
            key: ['key']
        },
        optional: {
            fileName: ['literal'],
            zone: ['literal'],
            subFolder: ['literal'],
            player: ['literal'],
            instance: ['boolean']
        }
    })
    const parseContents = validateContents<ParseAssetLegalContents>({
        contents,
        legalTags: ['Action', 'Computed', 'If', 'Exit', 'Feature', 'Bookmark', 'Import', 'Room', 'Variable', 'Map'],
        ignoreTags: ['Whitespace', 'Comment']
    })
    return {
        type: 'Tag',
        tag: {
            ...validate,
            tag: 'Story',
            startTagToken: open.startTagToken,
            endTagToken,
            contents: parseContents
        }
    }    
}

export default parseAssetFactory