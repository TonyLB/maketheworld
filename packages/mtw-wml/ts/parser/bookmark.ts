import { ParseBookmarkTag } from "./baseClasses"
import parseTaggedMessageContentsFactory from "./taggedMessage"

export const parseBookmarkFactory = parseTaggedMessageContentsFactory<ParseBookmarkTag>('Bookmark', ['String', 'Link', 'Bookmark', 'br', 'Space', 'If', 'Else', 'ElseIf'], { key: ['key'] })

export default parseBookmarkFactory
