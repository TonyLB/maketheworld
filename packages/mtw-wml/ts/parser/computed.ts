import { ConverterMixinFactory, isTypedParseTagOpen } from "../convert/functionMixins"
import { ParseTagFactory, ParseComputedTag } from "./baseClasses"
import { validateProperties, ExtractProperties, extractDependenciesFromJS } from "./utils"

export const parseComputedFactory: ParseTagFactory<ParseComputedTag> = ({ open, contents, endTagToken }) => {
    const validate = validateProperties<ExtractProperties<ParseComputedTag, never>>({
        open,
        endTagToken,
        required: {
            key: ['key'],
            src: ['expression']
        }
    })
    return {
        type: 'Tag',
        tag: {
            ...validate,
            tag: 'Computed',
            startTagToken: open.startTagToken,
            endTagToken,
            dependencies: extractDependenciesFromJS(validate.src)
        }
    }    
}

export const ParseComputedMixin = ConverterMixinFactory({
    typeGuard: isTypedParseTagOpen('Computed'),
    convert: ({ open, contents, endTagToken }) => {
        const validate = validateProperties<ExtractProperties<ParseComputedTag, never>>({
            open,
            endTagToken,
            required: {
                key: ['key'],
                src: ['expression']
            }
        })
        return {
            type: 'Tag',
            tag: {
                ...validate,
                tag: 'Computed',
                startTagToken: open.startTagToken,
                endTagToken,
                dependencies: extractDependenciesFromJS(validate.src)
            }
        }    
    }
})

export default parseComputedFactory
