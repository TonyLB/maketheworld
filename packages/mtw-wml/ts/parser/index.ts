import {
    Token,
    isStackTokenPropertyOrValue,
    TokenProperty,
    TokenExpressionValue,
    TokenKeyValue,
    TokenLiteralValue,
    isLegalBareToken,
    isTokenComment,
    isTokenWhitespace,
    isTokenDescription,
} from './tokenizer/baseClasses'

import {
    ParseTag,
    ParseTagFactory,
    ParseStackEntry,
    ParseStackTokenEntry,
    ParseException,
    ParseStackTagOpenEntry,
    isParseStackTagOpenEntry
} from './baseClasses'

import parseAssetFactory from './asset'
import parseRoomFactory from './room'
import parseFeatureFactory from './feature'
import parseConditionFactory from './condition'
import parseLinkFactory from './link'
import parseLineBreakFactory from './lineBreak'
import parseDescriptionFactory from './description'
import parseExitFactory from './exit'
import parseUseFactory from './use'
import parseImportFactory from './import'
import parseDependFactory from './depend'
import parseVariableFactory from './variable'
import parseComputedFactory from './computed'
import parseActionFactory from './action'
import parseNameFactory from './name'

export const createParseTag: ParseTagFactory<ParseTag> = (props) => {
    switch(props.open.tag) {
        case 'Asset':
            return parseAssetFactory(props)
        case 'Room':
            return parseRoomFactory(props)
        case 'Feature':
            return parseFeatureFactory(props)
        case 'Condition':
            return parseConditionFactory(props)
        case 'Link':
            return parseLinkFactory(props)
        case 'br':
            return parseLineBreakFactory(props)
        case 'Description':
            return parseDescriptionFactory(props)
        case 'Exit':
            return parseExitFactory(props)
        case 'Use':
            return parseUseFactory(props)
        case 'Import':
            return parseImportFactory(props)
        case 'Depend':
            return parseDependFactory(props)
        case 'Variable':
            return parseVariableFactory(props)
        case 'Computed':
            return parseComputedFactory(props)
        case 'Action':
            return parseActionFactory(props)
        case 'Name':
            return parseNameFactory(props)
        default:
            return {
                type: 'Tag',
                tag: {
                    tag: 'Comment',
                    startTagToken: props.open.startTagToken,
                    endTagToken: props.endTagToken
                }
            }
    }
}

export const parse = (tokens: Token[]): ParseTag[] => {
    let returnValue: ParseTag[] = []
    let stack: ParseStackEntry[] = []
    let context: ParseStackTagOpenEntry[] = []
    tokens.forEach((token, index) => {
        switch(token.type) {
            case 'TagOpenBegin':
                stack.push({
                    type: 'TagOpenPending',
                    tag: token.tag,
                    startTagToken: index
                })
                break
            case 'TagOpenEnd':
                let propertyTags: ParseStackTokenEntry<TokenProperty | TokenExpressionValue | TokenKeyValue | TokenLiteralValue>[] = []
                while(stack.length > 0) {
                    const stackItem = stack.pop()
                    if (stackItem.type === 'Tag') {
                        throw new ParseException('Unexpected tag-end', index, index)
                    }
                    if (stackItem.type === 'TagOpen') {
                        throw new ParseException('Illegal nested tag opening', index, index)
                    }
                    if (stackItem.type === 'Token') {
                        if (isStackTokenPropertyOrValue(stackItem)) {
                            propertyTags.push(stackItem)
                        }
                        continue
                    }
                    if (stackItem.type === 'TagOpenPending') {
                        //
                        // Assemble the tag open (possibly self-closing) from the open token, all
                        // the property tokens, and the close.
                        //
                        let properties: Record<string, (TokenExpressionValue | TokenKeyValue | TokenLiteralValue | boolean)> = {}
                        while(propertyTags.length > 0) {
                            const firstTag = propertyTags.pop()
                            if (firstTag.token.type !== 'Property') {
                                throw new ParseException('Unexpected value token', firstTag.index, firstTag.index)
                            }
                            //
                            // TODO: Validate that no key is assigned twice
                            //
                            if (firstTag.token.isBoolean) {
                                properties[firstTag.token.key] = firstTag.token.value
                            }
                            else {
                                const secondTag = propertyTags.pop()
                                if (secondTag.token.type === 'Property') {
                                    throw new ParseException('Unexpected property token', secondTag.index, secondTag.index)
                                }
                                else {
                                    properties[firstTag.token.key] = secondTag.token
                                }
                            }
                        }
                        if (token.selfClosing) {
                            const stackTag = createParseTag({
                                open: {
                                    type: 'TagOpen',
                                    tag: stackItem.tag,
                                    startTagToken: stackItem.startTagToken,
                                    properties
                                },
                                context: stack.filter(isParseStackTagOpenEntry),
                                contents: [],
                                endTagToken: index
                            })
                            if (stack.length > 0) {
                                stack.push(stackTag)
                            }
                            else {
                                returnValue.push(stackTag.tag)
                            }    
                        }
                        else {
                            stack.push({
                                type: 'TagOpen',
                                tag: stackItem.tag,
                                startTagToken: stackItem.startTagToken,
                                properties
                            })
                        }
                        break
                    }
                }
                break
            case 'TagClose':
                let contents: ParseTag[] = []
                while(stack.length > 0) {
                    const stackItem = stack.pop()
                    if (stackItem.type === 'TagOpenPending') {
                        throw new ParseException('Unexpected tag-close', index, index)
                    }
                    if (stackItem.type === 'Token') {
                        if (isLegalBareToken(stackItem.token)) {
                            if (isTokenComment(stackItem.token)) {
                                contents.unshift({
                                    tag: 'Comment',
                                    startTagToken: stackItem.index,
                                    endTagToken: stackItem.index
                                })
                            }
                            if (isTokenWhitespace(stackItem.token)) {
                                contents.unshift({
                                    tag: 'Whitespace',
                                    startTagToken: stackItem.index,
                                    endTagToken: stackItem.index
                                })
                            }
                            if (isTokenDescription(stackItem.token)) {
                                contents.unshift({
                                    tag: 'String',
                                    startTagToken: stackItem.index,
                                    endTagToken: stackItem.index,
                                    value: stackItem.token.value
                                })
                            }
                        }
                        continue
                    }
                    if (stackItem.type === 'Tag') {
                        contents.unshift(stackItem.tag)
                        continue
                    }
                    if (stackItem.type === 'TagOpen') {
                        const stackTag = createParseTag({
                            open: stackItem,
                            context: stack.filter(isParseStackTagOpenEntry),
                            contents,
                            endTagToken: index
                        })
                        if (stack.length > 0) {
                            stack.push(stackTag)
                        }
                        else {
                            returnValue.push(stackTag.tag)
                        }
                        break
                    }
                }
                break

            default:
                //
                // Ignore whitespace outside of tags
                //
                if (!isTokenWhitespace(token) || stack.length) {
                    stack.push({
                        type: 'Token',
                        token,
                        index
                    })
                }
        }
    })
    return returnValue
}

export default parse
