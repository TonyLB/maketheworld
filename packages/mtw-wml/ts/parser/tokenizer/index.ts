import {
    isTokenError,
    Token,
    TokenDescription,
    TokenTagClose,
    TokenTagOpenBegin,
    TokenTagOpenEnd,
    TokenWhitespace,
    TokenProperty,
    TokenBooleanProperty,
    TokenLiteralValue,
    TokenKeyValue,
    TokenExpressionValue
} from "./baseClasses"
import booleanPropertyTokenizer from "./boolean"
import expressionValueTokenizer from "./expression"
import keyValueTokenizer from "./key"
import literalValueTokenizer from "./literal"
import SourceStream from "./sourceStream"
import tagCloseTokenizer from "./tagClose"

import { tagOpenBeginTokenizer, tagOpenEndTokenizer } from "./tagOpen"
import tagPropertyTokenizer from "./tagProperty"
import whiteSpaceTokenizer from "./whiteSpace"
import { checkSubTokenizers } from "./utils"

enum TokenizerMode {
    Contents,
    Tag
}

export const tokenizer = (sourceStream: SourceStream): Token[] => {
    let mode: TokenizerMode = TokenizerMode.Contents
    let currentDescription: TokenDescription | undefined
    let returnValue: Token[] = []

    while(!sourceStream.isEndOfSource) {
        if (mode === TokenizerMode.Contents) {
            const checkSubTokens = checkSubTokenizers<TokenWhitespace | TokenTagClose | TokenTagOpenBegin>({
                sourceStream,
                subTokenizers: [whiteSpaceTokenizer, tagCloseTokenizer, tagOpenBeginTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenBegin') {
                        mode = TokenizerMode.Tag
                    }
                }
            })
            if (checkSubTokens) {
                if (checkSubTokens.success === false) {
                    return [checkSubTokens.error]
                }
                continue
            }
            return [{
                type: 'Error',
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                source: sourceStream.lookAhead(1),
                message: 'Unexpected token'
            }]
        }
        else {
            const checkSubTokens = checkSubTokenizers<TokenTagOpenEnd | TokenWhitespace | TokenProperty | TokenBooleanProperty>({
                sourceStream,
                subTokenizers: [tagOpenEndTokenizer, whiteSpaceTokenizer, tagPropertyTokenizer, booleanPropertyTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenEnd') {
                        mode = TokenizerMode.Contents
                    }
                    if (token.type === 'Property') {
                        //
                        // Consume any whitespace first...
                        //
                        const checkWhitespace = whiteSpaceTokenizer(sourceStream)
                        if (checkWhitespace) {
                            if (isTokenError(checkWhitespace)) {
                                return {
                                    success: false,
                                    error: checkWhitespace
                                }
                            }
                            returnValue.push(checkWhitespace)
                        }
                        //
                        // ... and then confirm that a value follows the key
                        //
                        const checkValue = checkSubTokenizers<TokenLiteralValue | TokenKeyValue | TokenExpressionValue>({
                            sourceStream,
                            subTokenizers: [literalValueTokenizer, keyValueTokenizer, expressionValueTokenizer],
                            callback: (token) => {
                                returnValue.push(token)
                            }
                        })
                        if (checkValue) {
                            if (checkValue.success === false) {
                                return checkValue
                            }
                        }
                        else {
                            return {
                                success: false,
                                error: {
                                    type: 'Error',
                                    source: sourceStream.lookAhead(1),
                                    startIdx: sourceStream.position,
                                    endIdx: sourceStream.position,
                                    message: 'Unexpected token'
                                }
                            }
                        }
                    }
                }
            })
            if (checkSubTokens) {
                if (checkSubTokens.success === false) {
                    return [checkSubTokens.error]
                }
                continue
            }
            return [{
                type: 'Error',
                source: sourceStream.lookAhead(1),
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                message: 'Unexpected token'
            }]
        }
    }
    return returnValue
}

export default tokenizer
