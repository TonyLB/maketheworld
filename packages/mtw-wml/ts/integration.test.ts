import {
    validatedSchema,
    schemaFromParse
} from '.'
import tokenizer from './parser/tokenizer'
import parser from './parser'
import SourceStream from './parser/tokenizer/sourceStream'
import wmlGrammar from './wmlGrammar/wml.ohm-bundle.js'

xdescribe('parallel validateSchema integration test (temporary)', () => {
    it('should correctly validate an asset schema', () => {
        const testSource = `
            <Asset key=(Test) fileName="test">
                <Import from=(BASE)>
                    <Use key=(basePower) type="Variable" as=(power) />
                    <Use key=(overview) type="Room" />
                </Import>
                <Room key=(ABC)>
                    <Name>Vortex</Name>
                    <Description>
                        Vortex
                        <Link to=(toggleOpen)>(toggle)</Link>
                    </Description>
                    <Exit from=(DEF)>vortex</Exit>
                </Room>
                <Condition if={open}>
                    <Depend on=(open) />
                    <Room key=(ABC)>
                        <Exit to=(DEF)>welcome</Exit>
                    </Room>
                </Condition>
                <Room key=(DEF)>
                    <Name>Welcome</Name>
                </Room>
                <Variable key=(open) default={false} />
                <Action key=(toggleOpen) src={open = !open} />
                <Computed key=(closed) src={!open}>
                    <Depend on=(open) />
                </Computed>
            </Asset>
        `
        const match = wmlGrammar.match(testSource)
        const schema = validatedSchema(match)
        const newSchema = schemaFromParse(parser(tokenizer(new SourceStream(testSource))))
        expect(newSchema[0]).toEqual(schema)
    })

    it('should correctly validate a character schema', () => {
        const testSource = `
            <Character key=(Tess) fileName="Tess" player="testy" zone="Library">
                <Image key=(icon) fileURL="testIcon.png" />
                <Name>Tess</Name>
                <Pronouns
                    subject="she"
                    object="her"
                    possessive="her"
                    adjective="hers"
                    reflexive="herself"
                />
            </Character>
        `
        const match = wmlGrammar.match(testSource)
        const schema = validatedSchema(match)
        const newSchema = schemaFromParse(parser(tokenizer(new SourceStream(testSource))))
        expect(newSchema[0]).toEqual(schema)
    })

    it('should correctly validate a map schema', () => {
        const testSource = `
            <Story key=(Test) instance fileName="test">
                <Map key=(TestMap)>
                    <Name>Test Map</Name>
                    <Image key=(ImageTest) fileURL="https://test.com/imageTest.png" />
                    <Room key=(ABC) x="200" y="150" />
                </Map>
            </Story>
        `
        const match = wmlGrammar.match(testSource)
        const schema = validatedSchema(match)
        const newSchema = schemaFromParse(parser(tokenizer(new SourceStream(testSource))))
        expect(newSchema[0]).toEqual(schema)
    })

})