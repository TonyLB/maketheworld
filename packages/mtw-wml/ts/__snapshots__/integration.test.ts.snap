// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize integration test (temporary) should correctly normalize an asset schema 1`] = `
Object {
  "ABC": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
        ],
        "display": undefined,
        "errors": Array [],
        "location": Array [
          0,
          1,
        ],
        "name": "Vortex",
        "props": Object {},
        "render": Array [
          Object {
            "spaceAfter": true,
            "spaceBefore": false,
            "tag": "String",
            "value": "Vortex",
          },
          Object {
            "contents": Array [],
            "props": Object {},
            "spaceAfter": false,
            "spaceBefore": true,
            "tag": "Link",
            "text": "(toggle)",
            "to": "toggleOpen",
          },
        ],
        "spaceAfter": false,
        "spaceBefore": false,
      },
      Object {
        "contents": Array [
          Object {
            "index": 0,
            "key": "ABC#DEF",
            "tag": "Exit",
          },
        ],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
          Object {
            "index": 0,
            "key": "Condition-0",
            "tag": "Condition",
          },
        ],
        "display": undefined,
        "errors": Array [],
        "location": Array [
          0,
          2,
          0,
        ],
        "props": Object {},
        "render": Array [],
        "spaceAfter": false,
        "spaceBefore": false,
      },
    ],
    "global": false,
    "key": "ABC",
    "tag": "Room",
  },
  "ABC#DEF": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
          Object {
            "index": 0,
            "key": "Condition-0",
            "tag": "Condition",
          },
          Object {
            "index": 1,
            "key": "ABC",
            "tag": "Room",
          },
        ],
        "errors": Array [],
        "location": Array [
          0,
          2,
          0,
          0,
        ],
        "props": Object {},
      },
    ],
    "from": "ABC",
    "key": "ABC#DEF",
    "name": "welcome",
    "tag": "Exit",
    "to": "DEF",
  },
  "Condition-0": Object {
    "appearances": Array [
      Object {
        "contents": Array [
          Object {
            "index": 1,
            "key": "ABC",
            "tag": "Room",
          },
        ],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
        ],
        "errors": Array [],
        "location": Array [
          0,
          2,
        ],
        "props": Object {},
      },
    ],
    "dependencies": Array [
      "open",
    ],
    "if": "open",
    "key": "Condition-0",
    "tag": "Condition",
  },
  "DEF": Object {
    "appearances": Array [
      Object {
        "contents": Array [
          Object {
            "index": 0,
            "key": "DEF#ABC",
            "tag": "Exit",
          },
        ],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
        ],
        "display": undefined,
        "errors": Array [],
        "location": Array [
          0,
          3,
        ],
        "name": "Welcome",
        "props": Object {},
        "render": undefined,
        "spaceAfter": false,
        "spaceBefore": false,
      },
    ],
    "global": false,
    "key": "DEF",
    "tag": "Room",
  },
  "DEF#ABC": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
          Object {
            "index": 0,
            "key": "DEF",
            "tag": "Room",
          },
        ],
        "errors": Array [],
        "location": Array [
          0,
          1,
          0,
        ],
        "props": Object {},
      },
    ],
    "from": "DEF",
    "key": "DEF#ABC",
    "name": "vortex",
    "tag": "Exit",
    "to": "ABC",
  },
  "Import-0": Object {
    "appearances": Array [
      Object {
        "contents": Array [
          Object {
            "index": 0,
            "key": "power",
            "tag": "Variable",
          },
          Object {
            "index": 0,
            "key": "overview",
            "tag": "Room",
          },
        ],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
        ],
        "errors": Array [],
        "location": Array [
          0,
          0,
        ],
        "props": Object {},
      },
    ],
    "from": "BASE",
    "key": "Import-0",
    "mapping": Object {
      "overview": Object {
        "key": "overview",
        "type": "Room",
      },
      "power": Object {
        "key": "basePower",
        "type": "Variable",
      },
    },
    "tag": "Import",
  },
  "Test": Object {
    "appearances": Array [
      Object {
        "contents": Array [
          Object {
            "index": 0,
            "key": "Import-0",
            "tag": "Import",
          },
          Object {
            "index": 0,
            "key": "ABC",
            "tag": "Room",
          },
          Object {
            "index": 0,
            "key": "DEF",
            "tag": "Room",
          },
          Object {
            "index": 0,
            "key": "Condition-0",
            "tag": "Condition",
          },
          Object {
            "index": 0,
            "key": "open",
            "tag": "Variable",
          },
          Object {
            "index": 0,
            "key": "toggleOpen",
            "tag": "Action",
          },
          Object {
            "index": 0,
            "key": "closed",
            "tag": "Computed",
          },
        ],
        "contextStack": Array [],
        "description": undefined,
        "errors": Array [],
        "location": Array [
          0,
        ],
        "props": Object {},
      },
    ],
    "fileName": "test",
    "key": "Test",
    "player": undefined,
    "tag": "Asset",
    "zone": undefined,
  },
  "closed": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
        ],
        "errors": Array [],
        "location": Array [
          0,
          6,
        ],
        "props": Object {},
      },
    ],
    "dependencies": Array [
      "open",
    ],
    "key": "closed",
    "src": "!open",
    "tag": "Computed",
  },
  "open": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
        ],
        "errors": Array [],
        "location": Array [
          0,
          4,
        ],
        "props": Object {},
      },
    ],
    "default": "false",
    "key": "open",
    "tag": "Variable",
  },
  "overview": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
          Object {
            "index": 0,
            "key": "Import-0",
            "tag": "Import",
          },
        ],
        "location": Array [
          0,
          0,
          1,
        ],
      },
    ],
    "key": "overview",
    "tag": "Room",
  },
  "power": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
          Object {
            "index": 0,
            "key": "Import-0",
            "tag": "Import",
          },
        ],
        "location": Array [
          0,
          0,
          0,
        ],
      },
    ],
    "key": "power",
    "tag": "Variable",
  },
  "toggleOpen": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
        ],
        "errors": Array [],
        "location": Array [
          0,
          5,
        ],
        "props": Object {},
      },
    ],
    "key": "toggleOpen",
    "src": "open = !open",
    "tag": "Action",
  },
}
`;

exports[`normalize integration test (temporary) should correctly validate a character schema 1`] = `
Object {
  "Tess": Object {
    "Name": "Tess",
    "Pronouns": Object {
      "adjective": "hers",
      "object": "her",
      "possessive": "her",
      "reflexive": "herself",
      "subject": "she",
    },
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [],
        "errors": Array [],
        "location": Array [
          0,
        ],
        "props": Object {},
      },
    ],
    "fileName": "Tess",
    "fileURL": "testIcon.png",
    "key": "Tess",
    "player": "testy",
    "tag": "Character",
    "zone": "Library",
  },
}
`;

exports[`normalize integration test (temporary) should correctly validate a map schema 1`] = `
Object {
  "ABC": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
          Object {
            "index": 0,
            "key": "TestMap",
            "tag": "Map",
          },
        ],
        "display": undefined,
        "errors": Array [],
        "location": Array [
          0,
          0,
          1,
        ],
        "props": Object {},
        "render": Array [],
        "spaceAfter": false,
        "spaceBefore": false,
      },
    ],
    "global": false,
    "key": "ABC",
    "tag": "Room",
  },
  "ImageTest": Object {
    "appearances": Array [
      Object {
        "contents": Array [],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
          Object {
            "index": 0,
            "key": "TestMap",
            "tag": "Map",
          },
        ],
        "errors": Array [],
        "location": Array [
          0,
          0,
          0,
        ],
        "props": Object {},
      },
    ],
    "fileURL": "https://test.com/imageTest.png",
    "key": "ImageTest",
    "tag": "Image",
  },
  "Test": Object {
    "Story": true,
    "appearances": Array [
      Object {
        "contents": Array [
          Object {
            "index": 0,
            "key": "TestMap",
            "tag": "Map",
          },
        ],
        "contextStack": Array [],
        "description": undefined,
        "errors": Array [],
        "location": Array [
          0,
        ],
        "props": Object {},
      },
    ],
    "fileName": "test",
    "instance": true,
    "key": "Test",
    "player": undefined,
    "tag": "Asset",
    "zone": undefined,
  },
  "TestMap": Object {
    "appearances": Array [
      Object {
        "contents": Array [
          Object {
            "index": 0,
            "key": "ImageTest",
            "tag": "Image",
          },
          Object {
            "index": 0,
            "key": "ABC",
            "tag": "Room",
          },
        ],
        "contextStack": Array [
          Object {
            "index": 0,
            "key": "Test",
            "tag": "Asset",
          },
        ],
        "errors": Array [],
        "images": Array [
          "ImageTest",
        ],
        "location": Array [
          0,
          0,
        ],
        "name": "Test Map",
        "props": Object {},
        "rooms": Object {
          "ABC": Object {
            "location": Array [
              0,
              0,
              1,
            ],
            "x": 200,
            "y": 150,
          },
        },
      },
    ],
    "key": "TestMap",
    "tag": "Map",
  },
}
`;
